lib/archethic/contracts.ex

```diff
@@ -40,98 +37,160 @@ defmodule Archethic.Contracts do
+         {:ok, calls} <- TransactionChain.fetch_contract_calls(previous_address) do
+      case execute_trigger(
+             trigger_type,
+             contract,
+             trigger_to_maybe_trigger_tx(trigger),
+             calls,
+             trigger_to_execute_opts(trigger)
+           ) do
+        {:ok, %Transaction{type: expected_type, data: expected_data}} ->
+          status == :tx_output &&
+            next_tx_type == expected_type &&
+            next_tx_data == expected_data
+
+        {:ok, nil} ->
+          status == :no_output
+
+        {:error, _} ->
+          status == :failure
+      end
     else
-      {:error, _} ->
+      _ ->
         false
     end
+  rescue
+    err ->
+      Logger.warn(Exception.format(:error, err, __STACKTRACE__))
+      false
   end
 
-  defp validate_conditions(version, conditions, prev_tx, next_tx) do
-    # here we only check that next_tx is allowed based on prev_tx inherit conditions
-    case Map.get(conditions, :inherit) do
+  # In the case of a trigger oracle/transaction,
+  # we need to fetch the transaction
+  defp trigger_to_maybe_trigger_tx({type, address}) when type in [:oracle, :transaction] do
+    {:ok, trigger_tx} = Archethic.search_transaction(address)
+    trigger_tx
+  end
+
+  defp trigger_to_maybe_trigger_tx(_), do: nil
+
+  # In the case of a trigger interval,
+  # because of the delay between execution and validation,
+  # we override the value returned by library function Time.now()
+  defp trigger_to_execute_opts({:interval, datetime}), do: [time_now: datetime]
+  defp trigger_to_execute_opts(_), do: []
+
+  @doc """
+  Validate any kind of condition.
+  The transaction and datetime depends on the condition.
+  """
+  @spec valid_condition?(
+          :oracle | :transaction | :inherit,
+          Contract.t(),
+          Transaction.t(),
+          DateTime.t()
+        ) :: boolean()
+  def valid_condition?(
+        condition_type,
+        contract = %Contract{version: version},
+        transaction = %Transaction{},
+        datetime
+      ) do
+    case get_condition(condition_type, contract) do
       nil ->
-        :ok
+        true
 
-      inherit_conditions ->
-        constants = %{
-          "previous" => Constants.from_transaction(prev_tx),
-          "next" => Constants.from_transaction(next_tx)
-        }
+      condition ->
+        constants = get_condition_constants(condition_type, contract, transaction, datetime)
+        Interpreter.valid_conditions?(version, condition, constants)
+    end
+  end
+
+  defp validate_trigger({:datetime, datetime}, _trigger, trigger_datetime) do
+    now = DateTime.utc_now()
 
-        if Interpreter.valid_conditions?(version, inherit_conditions, constants) do
-          :ok
-        else
-          Logger.error("Inherit constraints not respected")
-          {:error, :invalid_inherit_constraints}
-        end
+    # Accept time drifting for 10seconds
+    if DateTime.diff(trigger_datetime, datetime) >= 0 and
```

Comment:
Can this be refactored as it's used below in `validate_trigger({:interval, interval}` ?